### PR TITLE
Added DropConnect Layer and its tests.Fixes #413

### LINKED ISSUE #413: Add Dropout/DropConnect
Dropout is a recently introduced algorithm to prevent co-adaptation during training (overfitting). The key idea is to randomly drop units, along with their connections from a neural network during training. Routhly each element of a layer's output is kept with probability p, otherwise it's being set to 0.

For more information see:

* Geoffrey E. Hinton, Nitish Srivastava, Alex Krizhevsky, Ilya Sutskever, Ruslan R. Salakhutdinov, "Improving neural networks by preventing co-adaptation of feature detectors", 2012
* Nitish Srivastava, Geoffrey Hinton, Alex Krizhevsky, Ilya Sutskever, Ruslan Salakhutdinov, "Dropout: A Simple Way to Prevent Neural Networks from Overfitting", 2014

A simple way to implement the technique is to introduce a new function which creates a dropOutMask. Afterwards, we can multiply the dropOutMask with the inputActivation in all layers which should support dropout. Something like:

```
// dropout
if dropoutFraction > 0
{
	ActivationFunction::fn(inputActivation * dropOutMask, outputActivation);
}
else
{
	ActivationFunction::fn(inputActivation, outputActivation);
}
```
DropConnect is a generalization of Dropout that takes the idea a step further. Rather than zeroing each unit activations with probability p, it zeroes the weights/connections with probability p.

For more information see:

* Li Wan, Matthew Zeiler, Sixin Zhang, Yann Le Cun, Rob Fergus, "Regularization of Neural Networks using DropConnect", 2013

The idea to implement the technique is similar, except that we need to introduce the feature to all connections which should support dropConnect. The modified code should look something like:

```
// dropConnect
if dropConnectFraction > 0
{
	outputLayer.InputActivation() += (weights * dropOutMask * input);
}
else
{
	outputLayer.InputActivation() += (weights * input);
}
```

### PR DESCRIPTION
**Pull request by abhinavchanda**
DropConnect Layer is based on Linear Layer, but with the difference that the weight matrix is multiplied by the random mask.I have also added tests, in which , the network architechture and size of hidden layer is same as that of the dropout layer in tests/feedforward_network_test.cpp.On my systems, the three tests are giving these classificationErrors: 
  Dropout         DropConnect
1 0.0729288    0.0729288
2 0.02              0.02
3 0.134           0.068

### CODE REVIEW

**Comment by abhinavchanda** [comment]
@zoq I've added drop connect layer and also the tests. And the tests are passing on my system.Can you tell if any changes are required.

**Comment by palashahuja** [comment]
@abhinavchanda, The code that you have written is only for a linear layer. This needs to be generalized.

**Comment by zoq** [comment]
Merged DropConnect implementation; take a look at #576.

**Comment by abhinavchanda** [comment]
thanks for the reply.
